src/mumble/AudioInput.h

```diff
@@ -119,8 +119,18 @@ class Resynchronizer {
 	enum { S0, S1a, S1b, S2, S3, S4a, S4b, S5 } state = S0; ///< Queue fill control statemachine
 };
 
-class AudioInputRegistrar {
+struct EchoCancellationOption {
+	EchoCancellationOption(QString id, QString description, QString explanation) : id(id),
```

Comment:
The problem is different backends may offer different echo cancellation options. The purpose of this design is to let each backend handles its own options of echo cancellation. Creating an enum that would be shared by all backends will undo this effort.

I can also create en enum in each backend and convert them into int when saving into the Settings. Will that be okay?

Reply:
Ah okay I see.

> Creating an enum that would be shared by all backends will undo this effort.

How so? We could have backend-specific entries in that enum and then in the same way as in the current design not all backends support all possible Strings, the backends would not support all possible enum entries. I think that _should_ be possible :thinking:

Reply:
I don't want to make too many assumptions about what kind of echo cancellations we may have in the future, but if windows will have its own platform-supported echo cancellation, I suppose it should be "windows-echo" but not "mac-echo". It would be too fuzzy to call both "os-echo" since Mac's implementation may include acoustic hardware support. And the descriptive text displayed on the UI should also be different.

I think maybe I can define a series of int constants, or just an enum inside each audio backend and use the int constants or the enum in each backend to specify which echo cancellation is used, and store the int value into the Settings.

This way I think I can avoid using string but in the meantime keep the flexibility of this design. What do you think about this idea?

Reply:
I have tried the int constant in my newest commit. Would you have a look?